README.md
deluge gtk3
-----------

Import of Ubuntu's deluge 1.3.12 from `apt-source` with patches for GTK3 support.


debian/changelog
deluge (1.3.12-1ubuntu1+gtk3) xenial; urgency=medium

  * GTK3 support

 -- EiNSTeiN_ <einstein@g3nius.org>  Sat, 28 May 2016 14:00:00 -0500

deluge (1.3.12-1ubuntu1) wily; urgency=medium

  * Import new upstream release from Debian unstable.
    Remaining Ubuntu changes:
   - Explicitly recommend python-appindicator to ensure
     panel indicator is available (LP: #1384508).

 -- Andrew Starr-Bochicchio <a.starr.b@gmail.com>  Fri, 18 Sep 2015 22:28:47 -0400

deluge (1.3.12-1) unstable; urgency=medium

  * New upstream release.
  * Drop disable_SSLv3_rpcserver.patch, applied upstream.
  * Include full source copies of minified js files in
    the "debian/missing-sources" directory (Closes: #787550).
  * revert-erroneous-commit.patch: Revert erroneous fix
    backported from develop branch.

 -- Andrew Starr-Bochicchio <asb@debian.org>  Fri, 18 Sep 2015 21:34:24 -0400

deluge (1.3.10-3) unstable; urgency=medium

  * Fix permissions on /var/lib/deluged/config/ in
    deluged.postinst and properly remove all
    dpkg-statoverrides in deluged.postrm (Closes: #770628).

 -- Andrew Starr-Bochicchio <asb@debian.org>  Sun, 15 Feb 2015 11:14:44 -0500

deluge (1.3.10-2) unstable; urgency=high

  * debian/patches/disable_SSLv3_rpcserver.patch:
   - Disable SSLv3 for the RPCserver as Debian now
     builds OpenSSL without SSLv3 support altogether.
     Closes: #765765
  * Bump Standards-Version to 3.9.6, no changes needed.
  * Add full text of the CC-BY-3.0 to debian/copyright

 -- Andrew Starr-Bochicchio <asb@debian.org>  Sun, 26 Oct 2014 12:18:29 -0400

deluge (1.3.10-1) unstable; urgency=medium

  * New upstream bugfix release. 

 -- Andrew Starr-Bochicchio <asb@debian.org>  Wed, 15 Oct 2014 21:23:51 -0400

deluge (1.3.9-1) unstable; urgency=medium

  * New upstream release.
  * Use "canonical" URLs in the Vcs-* fields.
  * Update "Format" URL in debian/copyright.

 -- Andrew Starr-Bochicchio <asb@debian.org>  Sun, 05 Oct 2014 11:27:46 -0400

deluge (1.3.7-1) unstable; urgency=medium

  * New upstream release (Closes: #759217).
  * Add myself as an uploader.
  * Bump Standards-Version to 3.9.5, no changes needed.
  * Drop fix_plugins_tabs.patch, AutoExtractor_fix.patch,
    and Fix-twisted-13.1-compat.patch. All applied upstream.
  * deluged.postinst: Fix permissions on /var/lib/deluged
    Closes: #732596 Thanks to Diederik de Haas!

 -- Andrew Starr-Bochicchio <asb@debian.org>  Sat, 30 Aug 2014 20:20:56 -0700

deluge (1.3.6-1.1) unstable; urgency=medium

  * Non-maintainer upload.
  * Upload to unstable.
  * debian/patches/Fix-twisted-13.1-compat.patch: Backport
    upstream commits fixing Blocklist plugin incompatibility
    with twisted 13.1 (Closes: #729929). 
  * debian/patches/AutoExtractor_fix.patch: Fixes AutoExtractor
    Plugin failing to extract files with periods. Backported
    from upstream.
  * debian/patches/fix_plugins_tabs.patch: Fix issue with
    Plugins that add Tabs to torrentdetails. Backported from
    upstream.

 -- Andrew Starr-Bochicchio <asb@debian.org>  Mon, 17 Feb 2014 13:37:12 -0500

deluge (1.3.6-1) experimental; urgency=low

  * New upstream version.

 -- Cristian Greco <cristian@debian.org>  Wed, 19 Jun 2013 21:29:53 +0200

deluge (1.3.5-1) experimental; urgency=low

  * New upstream version. (Closes: #663250, #670366)
    - bring MainWindow to front when opening another instance.
      (Closes: #662733)
  * debian/deluged.postinst: check for already existing overrides
    before adding new ones. Thanks Andrew Starr-Bochicchio! (Closes: #671905)

 -- Cristian Greco <cristian@debian.org>  Tue, 08 May 2012 21:01:55 +0200

deluge (1.3.4-1) experimental; urgency=low

  [ Tanguy Ortolo ]
  * debian/deluged.init: initial work on init script for deluge daemon.

  [ Cristian Greco ]
  * New upstream version.
  * debian/deluged.init: some changes to configuration.
  * debian/control:
    - bump up Standards-Version to 3.9.3.
    - move transitional packages to oldlibs/extra.

 -- Cristian Greco <cristian@debian.org>  Sun, 11 Mar 2012 16:48:25 +0100

deluge (1.3.3-2) unstable; urgency=low

  * debian/patches:
    - new_release_check.patch: refreshed patch.
    - ubuntu_indicator.patch: new patch, enable indicator support
      by default on Ubuntu. (Closes: #641778) (LP: #843805)

 -- Cristian Greco <cristian@debian.org>  Sun, 25 Sep 2011 18:03:15 +0200

deluge (1.3.3-1) unstable; urgency=low

  * New upstream version.
  * debian/deluge-gtk.mime: provide MIME information for deluge-gtk.
  * debian/control:
    - add build-dep on intltool.
    - use my @debian.org address, drop DMUA field.
  * debian/rules: remove the .desktop file during clean.
  * debian/{control,rules}: dirty hack to avoid installing plugins eggs.
    Thanks Jakub Wilk!

 -- Cristian Greco <cristian@debian.org>  Thu, 25 Aug 2011 01:08:14 +0200

deluge (1.3.2-1) unstable; urgency=low

  * New upstream version.
    - fix arguments handling in deluge-console (Closes: #619565).
    - new_release_check.patch: refreshed.
  * debian/{control,rules}: switch from python-support to dh_python2.
  * debian/control: bump up Standards-Version to 3.9.2.
  * debian/control: move the python-libtorrent dependency from package
    deluge-common to packages deluged and deluge and clarify packages
    description, in order to be able to run the GTK+ user-interface as
    a thin client (LP: #672069).

 -- Cristian Greco <cristian@regolo.cc>  Wed, 08 Jun 2011 00:10:27 +0200

deluge (1.3.1-1) unstable; urgency=low

  * New upstream version.

 -- Cristian Greco <cristian@regolo.cc>  Wed, 03 Nov 2010 16:44:32 +0100

deluge (1.3.0-1) unstable; urgency=low

  * New upstream version.
    - remove all the stuff needed to track upstream git repository for
      Squeeze (get-orig-source in debian/control, debian/README.source
      and debian/patches/no_tag_build_dev.patch).
  * debian/patches/drop_deprecation_warnings.patch: deleted, it was a
    backport from upstream git repository.

deluge/ui/gtkui/peers_tab.py
#
# peers_tab.py
#
# Copyright (C) 2008 Andrew Resch <andrewresch@gmail.com>
#
# Deluge is free software.
#
# You may redistribute it and/or modify it under the terms of the
# GNU General Public License, as published by the Free Software
# Foundation; either version 3 of the License, or (at your option)
# any later version.
#
# deluge is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with deluge.    If not, write to:
#   The Free Software Foundation, Inc.,
#   51 Franklin Street, Fifth Floor
#   Boston, MA  02110-1301, USA.
#
#    In addition, as a special exception, the copyright holders give
#    permission to link the code of portions of this program with the OpenSSL
#    library.
#    You must obey the GNU General Public License in all respects for all of
#    the code used other than OpenSSL. If you modify file(s) with this
#    exception, you may extend this exception to your version of the file(s),
#    but you are not obligated to do so. If you do not wish to do so, delete
#    this exception statement from your version. If you delete this exception
#    statement from all source files in the program, then also delete it here.
#
#


from gi.repository import Gtk, GdkPixbuf
import os.path
import cPickle
import pkg_resources
from gi.repository import GObject
from itertools import izip

from deluge.ui.client import client
from deluge.configmanager import ConfigManager
import deluge.configmanager
import deluge.component as component
import deluge.common
from deluge.ui.gtkui.listview import cell_data_speed as cell_data_speed
from deluge.ui.gtkui.torrentdetails import Tab
from deluge.log import LOG as log
from deluge.ui.countries import COUNTRIES

def cell_data_progress(column, cell, model, row, data):
    value = model.get_value(row, data)
    cell.set_property("value", value * 100)
    cell.set_property("text", "%.2f%%" % (value * 100))

class PeersTab(Tab):
    def __init__(self):
        Tab.__init__(self)
        window = component.get("MainWindow")
        glade = window.get_glade()

        self._name = "Peers"
        self._child_widget = glade.get_object("peers_tab")
        self._tab_label = glade.get_object("peers_tab_label")
        self.peer_menu = glade.get_object("menu_peer_tab")
        window.insert_signals({
            "on_menuitem_add_peer_activate": self._on_menuitem_add_peer_activate,
            })

        self.listview = glade.get_object("peers_listview")
        self.listview.props.has_tooltip = True
        self.listview.connect("button-press-event", self._on_button_press_event)
        self.listview.connect("query-tooltip", self._on_query_tooltip)
        # country pixbuf, ip, client, downspeed, upspeed, country code, int_ip, seed/peer icon, progress
        self.liststore = Gtk.ListStore(GdkPixbuf.Pixbuf, str, str, int, int, str, float, GdkPixbuf.Pixbuf, float)
        self.cached_flag_pixbufs = {}

        self.seed_pixbuf = GdkPixbuf.Pixbuf.new_from_file(deluge.common.get_pixmap("seeding16.png"))
        self.peer_pixbuf = GdkPixbuf.Pixbuf.new_from_file(deluge.common.get_pixmap("downloading16.png"))

        # key is ip address, item is row iter
        self.peers = {}

        # Country column
        column = Gtk.TreeViewColumn()
        render = Gtk.CellRendererPixbuf()
        column.pack_start(render, False)
        column.add_attribute(render, "pixbuf", 0)
        column.set_sort_column_id(5)
        column.set_clickable(True)
        column.set_resizable(True)
        column.set_expand(False)
        column.set_min_width(20)
        column.set_reorderable(True)
        self.listview.append_column(column)

        # Address column
        column = Gtk.TreeViewColumn(_("Address"))
        render = Gtk.CellRendererPixbuf()
        column.pack_start(render, False)
        column.add_attribute(render, "pixbuf", 7)
        render = Gtk.CellRendererText()
        column.pack_start(render, False)
        column.add_attribute(render, "text", 1)
        column.set_sort_column_id(6)
        column.set_clickable(True)
        column.set_resizable(True)
        column.set_expand(False)
        column.set_min_width(100)
        column.set_reorderable(True)
        self.listview.append_column(column)

        # Client column
        column = Gtk.TreeViewColumn(_("Client"))
        render = Gtk.CellRendererText()
        column.pack_start(render, False)
        column.add_attribute(render, "text", 2)
        column.set_sort_column_id(2)
        column.set_clickable(True)
        column.set_resizable(True)
        column.set_expand(False)
        column.set_min_width(100)
        column.set_reorderable(True)
        self.listview.append_column(column)

        # Progress column
        column = Gtk.TreeViewColumn(_("Progress"))
        render = Gtk.CellRendererProgress()
        column.pack_start(render, True)
        column.set_cell_data_func(render, cell_data_progress, 8)
        column.set_sort_column_id(8)
        column.set_clickable(True)
        column.set_resizable(True)
        column.set_expand(False)
        column.set_min_width(100)
        column.set_reorderable(True)
        self.listview.append_column(column)

        # Down Speed column
        column = Gtk.TreeViewColumn(_("Down Speed"))
        render = Gtk.CellRendererText()
        column.pack_start(render, False)
        column.set_cell_data_func(render, cell_data_speed, 3)
        column.set_sort_column_id(3)
        column.set_clickable(True)
        column.set_resizable(True)
        column.set_expand(False)
        column.set_min_width(50)
        column.set_reorderable(True)
        self.listview.append_column(column)

        # Up Speed column
        column = Gtk.TreeViewColumn(_("Up Speed"))
        render = Gtk.CellRendererText()
        column.pack_start(render, False)
        column.set_cell_data_func(render, cell_data_speed, 4)
        column.set_sort_column_id(4)
        column.set_clickable(True)
        column.set_resizable(True)
        column.set_expand(False)
        column.set_min_width(50)
        # Bugfix: Last column needs max_width set to stop scrollbar appearing
        column.set_max_width(150)
        column.set_reorderable(True)
        self.listview.append_column(column)

        self.listview.set_model(self.liststore)

        self.load_state()

        self.torrent_id = None

    def save_state(self):
        filename = "peers_tab.state"
        # Get the current sort order of the view
        column_id, sort_order = self.liststore.get_sort_column_id()

        # Setup state dict
        state = {
            "columns": {},
            "sort_id": column_id,
            "sort_order": int(sort_order) if sort_order else None
        }

        for index, column in enumerate(self.listview.get_columns()):
            state["columns"][column.get_title()] = {
                "position": index,
                "width": column.get_width()
            }

        # Get the config location for saving the state file
        config_location = deluge.configmanager.get_config_dir()

        try:
            log.debug("Saving FilesTab state file: %s", filename)
            state_file = open(os.path.join(config_location, filename), "wb")
            cPickle.dump(state, state_file)
            state_file.close()
        except IOError, e:
            log.warning("Unable to save state file: %s", e)

    def load_state(self):
        filename = "peers_tab.state"
        # Get the config location for loading the state file
        config_location = deluge.configmanager.get_config_dir()
        state = None

        try:
            log.debug("Loading PeersTab state file: %s", filename)
            state_file = open(os.path.join(config_location, filename), "rb")
            state = cPickle.load(state_file)
            state_file.close()
        except (EOFError, IOError, AttributeError, cPickle.UnpicklingError), e:
            log.warning("Unable to load state file: %s", e)

        if state == None:
            return

        if len(state["columns"]) != len(self.listview.get_columns()):
            log.warning("peers_tab.state is not compatible! rejecting..")
            return

        if state["sort_id"] and state["sort_order"] is not None:
            self.liststore.set_sort_column_id(state["sort_id"], state["sort_order"])

        for (index, column) in enumerate(self.listview.get_columns()):
            cname = column.get_title()
            if state["columns"].has_key(cname):
                cstate = state["columns"][cname]
                column.set_sizing(Gtk.TreeViewColumnSizing.FIXED)
                column.set_fixed_width(cstate["width"] if cstate["width"] > 0 else 10)
                if state["sort_id"] == index and state["sort_order"] is not None:
                    column.set_sort_indicator(True)
                    column.set_sort_order(state["sort_order"])
                if cstate["position"] != index:
                    # Column is in wrong position
                    if cstate["position"] == 0:
                        self.listview.move_column_after(column, None)
                    elif self.listview.get_columns()[cstate["position"] - 1].get_title() != cname:
                        self.listview.move_column_after(column, self.listview.get_columns()[cstate["position"] - 1])

    def update(self):
        # Get the first selected torrent
        torrent_id = component.get("TorrentView").get_selected_torrents()

        # Only use the first torrent in the list or return if None selected
        if len(torrent_id) != 0:
            torrent_id = torrent_id[0]
        else:
            # No torrent is selected in the torrentview
            self.liststore.clear()
            return

        if torrent_id != self.torrent_id:
            # We only want to do this if the torrent_id has changed
            self.liststore.clear()
            self.peers = {}
            self.torrent_id = torrent_id

        component.get("SessionProxy").get_torrent_status(torrent_id, ["peers"]).addCallback(self._on_get_torrent_status)

    def get_flag_pixbuf(self, country):
        if country == "  ":
            return None

        if not self.cached_flag_pixbufs.has_key(country):
            # We haven't created a pixbuf for this country yet
            try:
                self.cached_flag_pixbufs[country] = GdkPixbuf.Pixbuf.new_from_file(
                    pkg_resources.resource_filename(
                        "deluge",
                         os.path.join("data", "pixmaps", "flags", country.lower() + ".png")))
            except Exception, e:
                log.debug("Unable to load flag: %s", e)
                return None

        return self.cached_flag_pixbufs[country]

    def _on_get_torrent_status(self, status):
        new_ips = set()
        for peer in status["peers"]:
            new_ips.add(peer["ip"])
            if self.peers.has_key(peer["ip"]):
                # We already have this peer in our list, so lets just update it
                row = self.peers[peer["ip"]]
                if not self.liststore.iter_is_valid(row):
                    # This iter is invalid, delete it and continue to next iteration
                    del self.peers[peer["ip"]]
                    continue
                values = self.liststore.get(row, 3, 4, 5, 7, 8)
                if peer["down_speed"] != values[0]:
                    self.liststore.set_value(row, 3, peer["down_speed"])
                if peer["up_speed"] != values[1]:
                    self.liststore.set_value(row, 4, peer["up_speed"])
                if peer["country"] != values[2]:
                    self.liststore.set_value(row, 5, peer["country"])
                    self.liststore.set_value(row, 0, self.get_flag_pixbuf(peer["country"]))
                if peer["seed"]:
                    icon = self.seed_pixbuf
                else:
                    icon = self.peer_pixbuf

                if icon != values[3]:
                    self.liststore.set_value(row, 7, icon)

                if peer["progress"] != values[4]:
                    self.liststore.set_value(row, 8, peer["progress"])
            else:
                # Peer is not in list so we need to add it

                # Create an int IP address for sorting purposes
                if peer["ip"].count(":") == 1:
                    # This is an IPv4 address
                    ip_int = sum([int(byte) << shift
                        for byte, shift in izip(peer["ip"].split(":")[0].split("."), (24, 16, 8, 0))])
                    peer_ip = peer["ip"]
                else:
                    # This is an IPv6 address
                    import socket
                    import binascii
                    # Split out the :port
                    ip = ":".join(peer["ip"].split(":")[:-1])
                    ip_int = long(binascii.hexlify(socket.inet_pton(socket.AF_INET6, ip)), 16)
                    peer_ip = "[%s]:%s" % (ip, peer["ip"].split(":")[-1])

                if peer["seed"]:
                    icon = self.seed_pixbuf
                else:
                    icon = self.peer_pixbuf

                row = self.liststore.append([
                    self.get_flag_pixbuf(peer["country"]),
                    peer_ip,
                    peer["client"],
                    peer["down_speed"],
                    peer["up_speed"],
                    peer["country"],
                    float(ip_int),
                    icon,
                    peer["progress"]])

                self.peers[peer["ip"]] = row

        # Now we need to remove any ips that were not in status["peers"] list
        for ip in set(self.peers.keys()).difference(new_ips):
            self.liststore.remove(self.peers[ip])
            del self.peers[ip]

    def clear(self):
        self.liststore.clear()

    def _on_button_press_event(self, widget, event):
        """This is a callback for showing the right-click context menu."""
        log.debug("on_button_press_event")
        # We only care about right-clicks
        if self.torrent_id and event.button == 3:
            self.peer_menu.popup(None, None, None, event.button, event.time)
            return True

    def _on_query_tooltip(self, widget, x, y, keyboard_tip, tooltip):
        context = widget.get_tooltip_context(x, y, keyboard_tip)
        (row_present, x, y, model, path, iter) = context

        if not row_present:
            return False
        else:
            country_code = model.get(iter, 5)[0]
            if country_code != "  " and country_code in COUNTRIES:
                tooltip.set_text(COUNTRIES[country_code])
                # widget here is self.listview
                widget.set_tooltip_cell(tooltip, path, widget.get_column(0),
                                        None)
                return True
            else:
                return False

    def _on_menuitem_add_peer_activate(self, menuitem):
        """This is a callback for manually adding a peer"""
        log.debug("on_menuitem_add_peer")
        dialog_glade = Gtk.Builder()
        dialog_glade.add_from_file(
            pkg_resources.resource_filename("deluge.ui.gtkui",
                "builder/dgtkpopups.ui"))
        peer_dialog = dialog_glade.get_object("connect_peer_dialog")
        txt_ip = dialog_glade.get_object("txt_ip")
        response = peer_dialog.run()
        if response:
            value = txt_ip.get_text()
            if value and ':' in value:
                if ']' in value:
                    #ipv6
                    ip = value.split("]")[0][1:]
                    port = value.split("]")[1][1:]
                else:
                    #ipv4
                    ip = value.split(":")[0]
                    port = value.split(":")[1]
                if deluge.common.is_ip(ip):
                    log.debug("adding peer %s to %s", value, self.torrent_id)
                    client.core.connect_peer(self.torrent_id, ip, port)
        peer_dialog.destroy()
        return True
